<!DOCTYPE html>
<html lang="en" dir="ltr">
	<head>
		<meta charset="utf-8" name="viewport" content="width=device-width, initial-scale=1.0" />
		<title>OpenVar Help</title>
		<link rel="stylesheet" href="/openvar/static/css/Doc_stylesheet.css" />
		<script src="https://code.jquery.com/jquery-3.5.1.min.js"></script>
		<script src="/openvar/static/js/OPV_help.js"></script>
	</head>
	
	<body>
		{% extends "OPV_navbar.html" %}
		{% block content %}
		<div class="doc__grid">
			
			<div class="doc__header">
				<h1> OpenVar Documentation </h1>
			</div>

			<div class="faq__principles">
				<h1> The hypotheses behind OpenVar </h1>
				<div class="faq_openvar">
					<p name="opv_q1" class="reveal_hide"> What is the genome annotation model used by OpenVar?</p>
					<p id="opv_q1" name="opv_q1" class="answer"> OpenVar uses the <a href="https://openprot.org/" title="OpenProt" target="_blank">OpenProt</a> annotation model. OpenProt is the first and only protein database to enforce a polycistronic model of mammalian genome annotation. Where common genome annotations report a single coding sequence (CDS) per transcript, OpenProt allows for multiple coding sequence per transcript. Therefore, OpenProt reports dual-coding and polycistronic genes in eukaryotic genomes. For more information, PMID <a href="https://pubmed.ncbi.nlm.nih.gov/33179748/" title="Brunet et al, NAR, 2019" target="_blank">33179748</a>. </p>
					<p name="opv_q2" class="reveal_hide"> What is OpenProt? How can I know more about OpenProt? </p>
					<p id="opv_q2" name="opv_q2" class="answer"> <a href="https://openprot.org/" title="OpenProt" target="_blank">OpenProt</a> is the first proteogenomic resource supporting a polycistronic annotation model of eukaryotic genomes. It is freely available at www.openprot.org. Every release and related articles can be found in the about section of the website or <a href="https://openprot.org/p/about" title="OpenProt About" target="_blank">here</a>. </p>
					<p name="opv_q3" class="reveal_hide"> What is an alternative open reading frame (altORF) / alternative protein (altProt)? </p>
					<p id="opv_q3" name="opv_q3" class="answer"> Current genome annotations in eukaryotes rely partly on ORF prediction algorithms, which are reliable only for sequences above a certain length. Consequently, three main criteria are enforced to distinguish true ORFs from randoms: (1) a minimum length of 100 codons; (2) a single coding sequence (CDS) per transcript; and (3) the use of an ATG start codon. However, in OpenProt, we use different terms to identify proteins based upon their genome annotation status:<br> - Proteins currently annotated in databases, such as UniProtKB, are translated from canonical CDSs (reference ORFs or refORFs) and are termed reference proteins (or RefProts). <br> - Alternative ORFs (or AltORFs) are defined as potential protein-coding ORF, located either in non-coding RNAs (e.g. long non-coding RNAs, pseudogene RNAs, etc...), or in UTRs or alternative reading frames overlapping a CDS in mRNAs. Predicted proteins translated from AltORFs are termed alternative proteins (or AltProts; IP_). AltProt and RefProt from a same gene are not isoforms: they are coded by different ORFs and their amino acid sequence is completely different. <br> - Predicted proteins translated from an alternative ORF (as defined above), but that either display (1) a close homology with a reference protein from the same gene; (2) the same start and/or stop codon than the reference protein; and an alignment score above the threshold are considered novel isoforms of the reference proteins (or Isoforms; II_). </p>
				</div>
			</div>

			<div class="faq__website">
				<h1> How to use the OpenVar website </h1>
				<div class="faq_website">
					<p name="web_q1" class="reveal_hide"> What do I need to submit an analysis to OpenVar? </p>
					<p id="web_q1" name="web_q1" class="answer"> To submit an analysis to OpenVar via the <a href="https://openprot.org/openvar/submit" title="Run Analysis" target="_blank">website</a>, you will need the folowwing:<br> - An email adress: the email adress will be used to send you a unique link to the results of the analysis.<br> - A study name: the study name will be used to identify your analysis report.<br> - The species: select the adequate species for your analysis.<br> - The genome assembly: select the adequate genome assembly used for variant calling (for Human, please select adequately between hg9, b37 and hg38).<br> - The desired genome build: select the genome build with which you would like to annotate your variant calling file (VCF).<br> - A variant calling file (VCF): upload your VCF. Please note that no VCFs are stored on our servers after your result link expires (10 days after completion of your analysis). Wait for completion of the upload to submit your analysis. </p>
					<p name="web_q2" class="reveal_hide"> When will I get my results? </p>
					<p id="web_q2" name="web_q2" class="answer"> This will depend on the size of the VCF you submit: for a VCF with 500 variants, expect 5 mins; for a VCF with 5 millions variants, expect 1 hour. </p>
					<p name="web_q3" class="reveal_hide"> Can I download the results of the analysis? What is in the output of the analysis? </p>
					<p id="web_q3" name="web_q3" class="answer"> Yes, absolutely! Everything that you see on the results page can be downloaded. You can download separately either your annotated VCF for downstream in house analysis, or the table of maximal impact annotation per ORF type for each variant. Alternatively, you can download everything as a zipped folder by clicking on "download all results".<br> The zipped folder contains the following:<br> - The submitted input vcf (input_vcf.vcf).<br> - A text file of all analysis warnings (warnings.txt ). This file would contain all SNPs that were not included in the analysis and the reason why (e.g. none of the alleles match the allele of the reference genome at the given position: invalid alleles).<br> - The usual annotated vcf (unique_id.ann.vcf).<br> - A tsv file listing all consequences for each variant (study_name_annOnePerLine.tsv). This file would thus contain as many lines as there are consequences for all of the submitted variants. For example, a SNP altering a canonical ORF and an alternative ORF would have 2 lines, one listing the effect on the canonical ORF, the second listing the effect on the alternative ORF.<br> - A tsv file listing only the maximal impact on canonical and alternative ORFs for each variant (study_name_max_impact.tsv). This file would thus contain as many lines as the number of submitted variants. For example, a SNP altering a canonical ORF and two alternative ORFs, the SNP line would contain the maximal impact for the canonical ORF alongside that of the highest impact on any of the alternative ORFs. Please note that the maximal impact is based on the SnpEff categorization in modifier, low, medium or high impact (for more information, you may want to consult the SnpEff documentation here). With this file, you always keep the effect on the canonical ORFs, but you also have the information gained from deeper annotation. In this file, two columns (ref_max_impact and alt_max_impact) are here to ease filtering in order to look at the variants with the highest impact in the alternative ORFs (alt_max_impact = 1). Please note that when the impact of the SNP is in the same category for the canonical ORF and the alternative ORF, both columns would be equal to 1 (ref_max_impact = 1 and alt_max_impact = 1).<br> - Figures: the zipped folder includes 4 figures which are in .svg format to easily customize and include in publications. The figures consists of the top 100 mutated genes (top_genes_var_rate.svg), the number of variants per category of impact in canonical and alternative proteins (var_per_impact.svg), the fold-change for each impact category gained with deeper annotation (impact_foldchange.svg), and a barchart of the number of altORF per the percentage of genes' variants clustered on the altORF (hotspots_barchart.svg). For the last one, for example at the far right of the graph, you would have the number of alternative proteins each explaining all of the SNPs of the corresponding gene. They are divided in categories based on the total number of SNPs: only one SNP, one to ten SNPs, and more than ten SNPs.<br> - A pickle object (summary.pkl) which contains all summary statistics needed to present the results on the web page in a python format. This should ease any downstream filtering you may want to perform.</p>
					<p name="web_q4" class="reveal_hide"> What are the impact categories reported by OpenVar? </p>
					<p id="web_q4" name="web_q4" class="answer"> OpenVar uses the same categorization of impacts as SNPEff, some examples are listed below:<br> - Modifier (impact 0): intergenic variant, intronic variants, and UTR variants.<br> - Low (impact 1): synonymous variants.<br> - Moderate (impact 2): missense variants.<br> - High (impact 3): insertion or deletion frameshift, premature stop codon.<br> For an exhaustive list of categorization of mutations, please consult the SNPEff documentation <a href="https://pcingola.github.io/SnpEff/se_inputoutput/#impact-prediction" title="impact prediction" target="_blank">here</a>. </p>
				</div>
			</div>
		</div>

		<div class="footer">
			<div>If you have any question, please <a href="https://openprot.org/p/ng/contactUs" title="Contact us" target="_self">contact us</a>.</div>
			<div><a href="https://twitter.com/intent/tweet?button_hashtag=OpenVar&ref_src=twsrc%5Etfw" class="twitter-hashtag-button" data-show-count="false">Tweet #OpenVar</a><script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script></div>
			<div>
				<p>OpenVar was developped by members of Pr Xavier Roucou lab:<br>Dr Marie A Brunet and Sebastien Leblanc.</p>
			</div>
			<div class="Sponsors">
				<img src="/openvar/static/images/CIHR_chair.png" alt="CIHR Research Chair">
				<img src="/openvar/static/images/Sherbrooke.jpg" alt="Universite de Sherbrooke">
				<img src="/openvar/static/images/CalculQuebec.png" alt="Calcul Quebec">
				<img src="/openvar/static/images/CalculCanada.png" alt="Calcul Canada">
			</div>
		</div>

		{% endblock %}
	</body>
</html>
