<!DOCTYPE html>
<html lang="en" dir="ltr">
	<head>
		<meta charset="utf-8" name="viewport" content="width=device-width, initial-scale=1.0" />
		<title>OpenVar Help</title>
		<link rel="stylesheet" href="/openvar/static/css/Doc_stylesheet.css" />
		<script src="https://code.jquery.com/jquery-3.5.1.min.js"></script>
		<script src="/openvar/static/js/OPV_help.js"></script>
	</head>
	
	<body>
		{% extends "OPV_navbar.html" %}
		{% block content %}
		<div class="doc__grid">
			
			<div class="doc__header">
				<h1> OpenVar Documentation </h1>
			</div>

			<div class="faq__principles">
				<h1> The hypotheses behind OpenVar </h1>
				<div class="faq_openvar">
					<p name="opv_q1" class="reveal_hide"> What is the genome annotation model used by OpenVar?</p>
					<p id="opv_q1" name="opv_q1" class="answer"> OpenVar uses the <a href="https://openprot.org/" title="OpenProt" target="_blank">OpenProt</a> annotation model. OpenProt is the first and only protein database to enforce a polycistronic model of mammalian genome annotation. Where common genome annotations report a single coding sequence (CDS) per transcript, OpenProt allows for multiple coding sequence per transcript. Therefore, OpenProt reports dual-coding and polycistronic genes in eukaryotic genomes. For more information, PMID <a href="https://pubmed.ncbi.nlm.nih.gov/33179748/" title="Brunet et al, NAR, 2019" target="_blank">33179748</a>. </p>
					<p name="opv_q2" class="reveal_hide"> What is OpenProt? How can I know more about OpenProt? </p>
					<p id="opv_q2" name="opv_q2" class="answer"> <a href="https://openprot.org/" title="OpenProt" target="_blank">OpenProt</a> is the first proteogenomic resource supporting a polycistronic annotation model of eukaryotic genomes. It is freely available at www.openprot.org. Every release and related articles can be found in the about section of the website or <a href="https://openprot.org/p/about" title="OpenProt About" target="_blank">here</a>. </p>
					<p name="opv_q3" class="reveal_hide"> What is an alternative open reading frame (altORF) / alternative protein (altProt)? </p>
					<p id="opv_q3" name="opv_q3" class="answer"> Current genome annotations in eukaryotes rely partly on ORF prediction algorithms, which are reliable only for sequences above a certain length. Consequently, three main criteria are enforced to distinguish true ORFs from randoms: (1) a minimum length of 100 codons; (2) a single coding sequence (CDS) per transcript; and (3) the use of an ATG start codon. However, in OpenProt, we use different terms to identify proteins based upon their genome annotation status.<br> Proteins currently annotated in databases, such as UniProtKB, are translated from canonical CDSs (reference ORFs or refORFs) and are termed reference proteins (or RefProts). <br> Alternative ORFs (or AltORFs) are defined as potential protein-coding ORF, located either in non-coding RNAs (e.g. long non-coding RNAs, pseudogene RNAs, etc...), or in UTRs or alternative reading frames overlapping a CDS in mRNAs. Predicted proteins translated from AltORFs are termed alternative proteins (or AltProts; IP_). AltProt and RefProt from a same gene are not isoforms: they are coded by different ORFs and their amino acid sequence is completely different. <br> Predicted proteins translated from an alternative ORF (as defined above), but that either display (1) a close homology with a reference protein from the same gene; (2) the same start and/or stop codon than the reference protein; and an alignment score above the threshold are considered novel isoforms of the reference proteins (or Isoforms; II_). </p>
				</div>
			</div>

			<div class="faq__website">
				<h1> How to use the OpenVar website </h1>
				<div class="faq_website">
					<p name="web_q1" class="reveal_hide"> This is the first question </p>
					<p id="web_q1" name="web_q1" class="answer"> This is the first response </p>
					<p name="web_q2" class="reveal_hide"> This is the second question </p>
					<p id="web_q2" name="web_q2" class="answer"> This is the second response </p>
				</div>
			</div>
		</div>

		<div class="footer">
			<div>If you have any question, please <a href="https://openprot.org/p/ng/contactUs" title="Contact us" target="_self">contact us</a>.</div>
			<div><a href="https://twitter.com/intent/tweet?button_hashtag=OpenVar&ref_src=twsrc%5Etfw" class="twitter-hashtag-button" data-show-count="false">Tweet #OpenVar</a><script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script></div>
			<div>
				<p>OpenVar was developped by members of Pr Xavier Roucou lab:<br>Dr Marie A Brunet and Sebastien Leblanc.</p>
			</div>
			<div class="Sponsors">
				<img src="/openvar/static/images/CIHR_chair.png" alt="CIHR Research Chair">
				<img src="/openvar/static/images/Sherbrooke.jpg" alt="Universite de Sherbrooke">
				<img src="/openvar/static/images/CalculQuebec.png" alt="Calcul Quebec">
				<img src="/openvar/static/images/CalculCanada.png" alt="Calcul Canada">
			</div>
		</div>

		{% endblock %}
	</body>
</html>
